composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,235 @@
+        #    ...
+        # ]
+
+        # Contains 1 / d^2 of each station
+        factor_list = []
+        # The total sum of 1 / d^2 of all the stations
+        factor_sum = 0.0
+        for row in input_list:
+            latitude = row[1]
+            longitude = row[2]
+            # Calculate the distance from each station to Phoenix
+            distance_to_phx = math.sqrt(
+                (phx_location[0] - latitude) ** 2 + (phx_location[1] - longitude) ** 2
+            )
+            # Calculate the distance factor of each station (1 / d^2)
+            distance_factor = 1.0 / (distance_to_phx**2)
+            factor_list.append(distance_factor)
+            factor_sum += distance_factor
+
+        # Contains the weights of each station
+        weights_list = []
+        for val in factor_list:
+            weights_list.append(val / factor_sum)
+
+        dwa_result = 0.0
+        for row in input_list:
+            # This is the annual prcipitation/snowfall of each station
+            annual_value = row[4]
+            # Weight of each station
+            weight = weights_list[input_list.index(row)]
+            dwa_result += weight * annual_value
+
+        return dwa_result
+
+    for year_val in range(1997, 2022):
+        # Collect() function returns a list of Row object.
+        # prcp_year and snow_year will be the input for the distance weighting algorithm
+        prcp_year = (
+            annual_df.where(
+                (annual_df.ELEMENT == "PRCP") & (annual_df.YEAR == year_val)
+            )
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_PRCP"))
+            .collect()
+        )
+
+        snow_year = (
+            annual_df.where(
+                (annual_df.ELEMENT == "SNOW") & (annual_df.YEAR == year_val)
+            )
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_SNOW"))
+            .collect()
+        )
+
+        phx_annual_prcp_df = phx_annual_prcp_df.withColumn(
+            f"PHX_PRCP_{year_val}", lit(phx_dw_compute(prcp_year))
+        )
+        phx_annual_snow_df = phx_annual_snow_df.withColumn(
+            f"PHX_SNOW_{year_val}", lit(phx_dw_compute(snow_year))
+        )
+
+    # This table has only two rows (the first row contains the columns)
+    phx_annual_prcp_df.show()
+    phx_annual_snow_df.show()
+
+    # Write results to GCS
+    if "--dry-run" in sys.argv:
+        print("Data will not be uploaded to BigQuery")
+    else:
+        # Set GCS temp location
+        temp_path = BUCKET_NAME
+
+        # Saving the data to BigQuery using the "indirect path" method and the spark-bigquery connector
+        # Uses the "overwrite" SaveMode to ensure DAG doesn't fail when being re-run
+        # See https://spark.apache.org/docs/latest/sql-data-sources-load-save-functions.html#save-modes
+        # for other save mode options
+        (
```

Comment:
You can remove the need for a staging GCS bucket using the `direct` write method: https://github.com/GoogleCloudDataproc/spark-bigquery-connector#direct-write-using-the-bigquery-storage-write-api-preview

Reply:
will do